### ISSUE TITLE
RS1016 is raised on abstract class instead of on inheritors

### ISSUE THREAD

**Issue by reduckted** [dotnet/roslyn-analyzers#5818]
### Analyzer

**Diagnostic ID**: [RS1016](https://github.com/dotnet/roslyn-analyzers/blob/ffd024eb265e97e639a2879d869cf82b5b958992/src/Microsoft.CodeAnalysis.Analyzers/Core/FixAnalyzers/FixerWithFixAllAnalyzer.cs): `Code fix providers should provide FixAll support`

### Analyzer source

**NuGet Package**: [Microsoft.CodeAnalysis.Analyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.Analyzers/3.3.3)

**Version**: 3.3.3

### Describe the bug

When you define an abstract class that inherits from `CodeFixProvider` and _do not_ override the `GetfixAllProvider()` method, RS1016 is raised on the abstract class instead of inheritors.

### Steps To Reproduce

Compile this code:

```csharp
using System.Collections.Immutable;
using System.Threading.Tasks;
using Microsoft.CodeAnalysis.CodeActions;
using Microsoft.CodeAnalysis.CodeFixes;

namespace Repro;

public abstract class CodeFixProviderBase : CodeFixProvider
{
    // Contains common helper methods for inheritors.
}

public class MyAnalyzer : CodeFixProviderBase
{
    public override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create("FOO001");

    public override Task RegisterCodeFixesAsync(CodeFixContext context)
    {
        context.RegisterCodeFix(
            CodeAction.Create("Foo",(cancellation) => { return Task.FromResult(context.Document); }),
            context.Diagnostics[0]
        );

        return Task.CompletedTask;
    }
}
```

### Expected behavior

RS1016 is produced on the `MyAnalyzer` class.

### Actual behavior

RS1016 is produced on the `CodeFixProviderBase` class.

## Additional context

Even when the inheriting class overrides `GetFixAllProvider`, the RS1016 diagnostic is still reported on the base class, even though there's no need for it to provide a "fix all" provider. For example, this still produces RS1016:

```csharp
public abstract class CodeFixProviderBase : CodeFixProvider
                   // ~~~~~~~~~~~~~~~~~~~ RS1016
{
    // Contains common helper methods for inheritors.
}

public class MyAnalyzer : CodeFixProviderBase
{
    public override FixAllProvider GetFixAllProvider() => WellKnownFixAllProviders.BatchFixer;

    // Rest of the code fix provider implementation...
}
```

This can be worked around by overriding `GetFixAllProvider()` in `CodeFixProviderBase` and marking it as abstract. This forces inheritors to override it. but I feel that it would be better for the diagnostic to be recorded on the inheritors instead of the base class.

**Comment by Youssef1313** [comment]
https://github.com/dotnet/roslyn-analyzers/blob/f471d3381584f10f9908432e0b2b2b8ef07a0aa6/src/Microsoft.CodeAnalysis.Analyzers/Core/FixAnalyzers/FixerWithFixAllAnalyzer.cs#L243

We may rely on `ExportCodeFixProvider` for this condition instead of comparing the base type with `CodeFixProvider`
